pkg/kubevirt/utils.go

```diff
@@ -138,36 +146,86 @@ func buildVirtualMachineInstanceTemplate(ctx *context.MachineContext) *kubevirtv
 
 	template.Spec = *ctx.KubevirtMachine.Spec.VirtualMachineTemplate.Spec.Template.Spec.DeepCopy()
 
-	cloudInitVolumeName := "cloudinitvolume"
-	cloudInitVolume := kubevirtv1.Volume{
-		Name: cloudInitVolumeName,
-		VolumeSource: kubevirtv1.VolumeSource{
-			CloudInitConfigDrive: &kubevirtv1.CloudInitConfigDriveSource{
-				UserDataSecretRef: &corev1.LocalObjectReference{
-					Name: *ctx.Machine.Spec.Bootstrap.DataSecretName + "-userdata",
+	cloudInitType, index := detectCloudInitVolumeType(template)
+	cloudInitVolume := cloudinitVolume(template, ctx, cloudInitType)
+	cloudInitDisk := cloudinitDisk(template, cloudInitType)
+	switch cloudInitType {
+	case "CloudInitNoCloud":
+		err := mergo.Merge(&template.Spec.Volumes[index], cloudInitVolume)
+		return nil, err
+	case "CloudInitConfigDrive":
+		template.Spec.Volumes = append(template.Spec.Volumes, cloudInitVolume)
+	}
+	if found, _ := detectCloudInitDisk(template); !found {
+		template.Spec.Domain.Devices.Disks = append(template.Spec.Domain.Devices.Disks, cloudInitDisk)
+	}
+
+	return template, nil
+}
+
+// nodeRole returns the role of this node ("control-plane" or "worker").
+func nodeRole(ctx *context.MachineContext) string {
+	if util.IsControlPlaneMachine(ctx.Machine) {
+		return constants.ControlPlaneNodeRoleValue
+	}
+	return constants.WorkerNodeRoleValue
+}
+
+func cloudinitVolume(vmi *kubevirtv1.VirtualMachineInstanceTemplateSpec, ctx *context.MachineContext, preferredType string) kubevirtv1.Volume {
+	switch preferredType {
+	case "CloudInitNoCloud":
+		return kubevirtv1.Volume{
+			Name: cloudInitVolumeName,
+			VolumeSource: kubevirtv1.VolumeSource{
+				CloudInitNoCloud: &kubevirtv1.CloudInitNoCloudSource{
+					UserDataSecretRef: &corev1.LocalObjectReference{
+						Name: *ctx.Machine.Spec.Bootstrap.DataSecretName + "-userdata",
+					},
 				},
 			},
-		},
+		}
+	default:
+		return kubevirtv1.Volume{
+			Name: cloudInitVolumeName,
+			VolumeSource: kubevirtv1.VolumeSource{
+				CloudInitConfigDrive: &kubevirtv1.CloudInitConfigDriveSource{
+					UserDataSecretRef: &corev1.LocalObjectReference{
+						Name: *ctx.Machine.Spec.Bootstrap.DataSecretName + "-userdata",
+					},
+				},
+			},
+		}
```

Comment:
Not sure about this function. Another way to implement this logic will be to split it to two functions, each generating one version of cloud init volume. Then use each one of them in the matching case in the switch sentence in the `buildVirtualMachineInstanceTemplate` function above, where we already know the volume type.

The reason I'm not comfortable with the current implementation is that we ask the same question again and again (what is the volume type?), while we already know the answer.